sim/hunter/talents.go

```diff
@@ -1,870 +1,823 @@
+			if spell == hunter.ExplosiveShot {
+				aura.RemoveStack(sim)
+			}
+		},
+	})
+
+	hunter.RegisterAura(core.Aura{
+		Label:    "TNT Talent",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+		OnPeriodicDamageDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			if spell != hunter.BlackArrow && spell != hunter.ExplosiveTrap {
+				return
+			}
+
+			if !icd.IsReady(sim) {
+				return
+			}
+
+			if sim.RandomFloat("Lock and Load") < procChance {
+				icd.Use(sim)
+				hunter.LockAndLoadAura.Activate(sim)
+				hunter.LockAndLoadAura.SetStacks(sim, 2)
+			}
+		},
+	})
+}
+
+func (hunter *Hunter) applyThrillOfTheHunt() {
+	if hunter.Talents.ThrillOfTheHunt == 0 {
+		return
+	}
+
+	procChance := float64(hunter.Talents.ThrillOfTheHunt) * 0.05
+	focusMetrics := hunter.NewFocusMetrics(core.ActionID{SpellID: 34499})
+
+	hunter.RegisterAura(core.Aura{
+		Label:    "Thrill of the Hunt",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+		OnSpellHitDealt: func(aura *core.Aura, sim *core.Simulation, spell *core.Spell, result *core.SpellResult) {
+			// mask 256
+			if spell != hunter.ArcaneShot || spell != hunter.ExplosiveShot || spell != hunter.BlackArrow {
+				return
+			}
+
+			if sim.Proc(procChance, "ThrillOfTheHunt") {
+				hunter.AddFocus(sim, spell.CurCast.Cost*0.4, focusMetrics)
+			}
+		},
+	})
+}
+
+func (hunter *Hunter) applySniperTraining() {
+	if hunter.Talents.SniperTraining == 0 {
+		return
+	}
+
+	uptime := hunter.SurvivalOptions.SniperTrainingUptime
+	if uptime <= 0 {
+		return
+	}
+	uptime = min(1, uptime)
+
+	dmgMod := .02 * float64(hunter.Talents.SniperTraining)
+
+	stAura := hunter.RegisterAura(core.Aura{
+		Label:    "Sniper Training",
+		ActionID: core.ActionID{SpellID: 53304},
+		Duration: time.Second * 15,
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			if hunter.SteadyShot != nil {
+				hunter.SteadyShot.DamageMultiplierAdditive += dmgMod
+			}
+			if hunter.CobraShot != nil {
+				hunter.CobraShot.DamageMultiplierAdditive += dmgMod
+			}
+		},
+		OnExpire: func(aura *core.Aura, sim *core.Simulation) {
+			if hunter.SteadyShot != nil {
+				hunter.SteadyShot.DamageMultiplierAdditive -= dmgMod
+			}
+			if hunter.CobraShot != nil {
+				hunter.CobraShot.DamageMultiplierAdditive -= dmgMod
+			}
+		},
+	})
+
+	core.ApplyFixedUptimeAura(stAura, uptime, time.Second*15, 1)
+}
+
+func (hunter *Hunter) applyHuntingParty() {
+	if !hunter.Talents.HuntingParty {
+		return
+	}
+
+	hunter.RegisterAura(core.Aura{
+		Label:    "Hunting Party",
+		Duration: core.NeverExpires,
+		OnReset: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Activate(sim)
+		},
+		OnGain: func(aura *core.Aura, sim *core.Simulation) {
+			aura.Unit.MultiplyAttackSpeed(sim, 1.10)
+		},
```

Comment:
Missing OnExpire which would casuse this aura to constantly stack 10% attack speed on every single iteration